// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: third_party/xkratos/api/paginator/paginator.proto

package paginator

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on PaginatorReq with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *PaginatorReq) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on PaginatorReq with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in PaginatorReqMultiError, or
// nil if none found.
func (m *PaginatorReq) ValidateAll() error {
	return m.validate(true)
}

func (m *PaginatorReq) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetPerPage() <= 0 {
		err := PaginatorReqValidationError{
			field:  "PerPage",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if m.GetCurrentPage() <= 0 {
		err := PaginatorReqValidationError{
			field:  "CurrentPage",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if m.GetOrder() != "" {

		if l := utf8.RuneCountInString(m.GetOrder()); l < 1 || l > 20 {
			err := PaginatorReqValidationError{
				field:  "Order",
				reason: "value length must be between 1 and 20 runes, inclusive",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

	}

	if len(errors) > 0 {
		return PaginatorReqMultiError(errors)
	}

	return nil
}

// PaginatorReqMultiError is an error wrapping multiple validation errors
// returned by PaginatorReq.ValidateAll() if the designated constraints aren't met.
type PaginatorReqMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m PaginatorReqMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m PaginatorReqMultiError) AllErrors() []error { return m }

// PaginatorReqValidationError is the validation error returned by
// PaginatorReq.Validate if the designated constraints aren't met.
type PaginatorReqValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e PaginatorReqValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e PaginatorReqValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e PaginatorReqValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e PaginatorReqValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e PaginatorReqValidationError) ErrorName() string { return "PaginatorReqValidationError" }

// Error satisfies the builtin error interface
func (e PaginatorReqValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sPaginatorReq.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = PaginatorReqValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = PaginatorReqValidationError{}

// Validate checks the field values on PaginatorReply with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *PaginatorReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on PaginatorReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in PaginatorReplyMultiError,
// or nil if none found.
func (m *PaginatorReply) ValidateAll() error {
	return m.validate(true)
}

func (m *PaginatorReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for ItemsCount

	// no validation rules for PagesCount

	if len(errors) > 0 {
		return PaginatorReplyMultiError(errors)
	}

	return nil
}

// PaginatorReplyMultiError is an error wrapping multiple validation errors
// returned by PaginatorReply.ValidateAll() if the designated constraints
// aren't met.
type PaginatorReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m PaginatorReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m PaginatorReplyMultiError) AllErrors() []error { return m }

// PaginatorReplyValidationError is the validation error returned by
// PaginatorReply.Validate if the designated constraints aren't met.
type PaginatorReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e PaginatorReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e PaginatorReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e PaginatorReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e PaginatorReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e PaginatorReplyValidationError) ErrorName() string { return "PaginatorReplyValidationError" }

// Error satisfies the builtin error interface
func (e PaginatorReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sPaginatorReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = PaginatorReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = PaginatorReplyValidationError{}
